crates/app/src/values.rs

```diff
@@ -0,0 +1,57 @@
+use serde::Serialize;
+use serde_json::Value;
+
+// ValuesMap stores dynamically-typed values.
+pub type ValuesMap = serde_json::Map<String, Value>;
+
+/// ValuesMapBuilder assists in building a ValuesMap.
+#[derive(Default)]
+pub struct ValuesMapBuilder(ValuesMap);
+
+impl ValuesMapBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn string(&mut self, key: impl Into<String>, value: impl Into<String>) -> &mut Self {
+        self.entry(key, value.into())
+    }
+
+    pub fn string_option(
+        &mut self,
+        key: impl Into<String>,
+        value: Option<impl Into<String>>,
+    ) -> &mut Self {
+        if let Some(value) = value {
+            self.0.insert(key.into(), value.into().into());
+        }
+        self
+    }
+
+    pub fn string_array<T: Into<String>>(
+        &mut self,
+        key: impl Into<String>,
+        iter: impl IntoIterator<Item = T>,
+    ) -> &mut Self {
+        self.entry(key, iter.into_iter().map(|s| s.into()).collect::<Vec<_>>())
+    }
+
+    pub fn entry(&mut self, key: impl Into<String>, value: impl Into<Value>) -> &mut Self {
+        self.0.insert(key.into(), value.into());
+        self
+    }
+
+    pub fn serializable(
+        &mut self,
+        key: impl Into<String>,
+        value: impl Serialize,
+    ) -> serde_json::Result<&mut Self> {
+        let value = serde_json::to_value(value)?;
+        self.0.insert(key.into(), value);
+        Ok(self)
+    }
+
+    pub fn build(&mut self) -> ValuesMap {
```

Comment:
Minor nit: I generally expect the `build` method of a builder to take ownership of `self`, which makes it obvious that I can't use the builder anymore.  Here I see I can reuse the builder, but that's not super useful given that it's been cleared out, plus I wouldn't know it is cleared out without looking at the code.

So I'd like to see this either take ownership of `self` or a doc comment describing the behavior.

Reply:
I moved this behavior into a new `take` method and updated `build` to consume self.